xdsl/ir.py

```diff
@@ -309,11 +309,25 @@ def print_parameter(data: DataElement, printer: Printer) -> None:
         """Print the attribute parameter."""
 
 
+_PA = TypeVar("_PA", bound="ParametrizedAttribute")
+
+
 @dataclass(frozen=True)
 class ParametrizedAttribute(Attribute):
     """An attribute parametrized by other attributes."""
     parameters: list[Attribute] = field(default_factory=list)
 
+    @classmethod
+    def new(cls: type[_PA], params: list[Attribute]) -> _PA:
+        # Create the new attribute object, without calling its __init__.
+        # We do this to allow users to redefine their own __init__.
+        attr = cls.__new__(cls)
```

Comment:
Okay, I actually NEVER thought this way of building classes would come in clutch at a point in my life. But here it is. Well played, Python, you win this round, well played!

Reply:
Why exactly are we instantiating attributes without calling their constructor here? Is it so we can use this as a generic way to create them when parsing? It feels like this might lead to some counter-intuitive and somewhat hard to explain user errors where they expect the constructor to be called, but it isn't...

A part of me wants to keep the constructor reserved for being the actual way the object is constructed, but I understand that it would make the thing less usable, and introduces `.get()`ers, which I find are also really ugly and counter-intuitive.

I can't really think of a better way of doing it. We might just have to accept this and keep it in mind for later.

Reply:
The issue with calling the normal constructor is that we want to give users the option to override that exact thing. So, by dynamic dispatch, we would have pretty much unpredictable behavior in our code.

I think this is a very cool way of doing it. I expect users to define their own `__init__` functions, if they aren't happy with whatever dataclass provides.

Reply:
The core differences between us and actual python is that python just instantiates (through init), whereas we want to instantiate, change fields, run verifiers, and so on and then give the operation back to the user. Therefore, we cannot use jsut `__init__`.

Reply:
A proposal I had before was letting users define operation factories instead of the actual operations. This could solve the issue here as well, as we could then just instantiate the operations and change the fields/run verifiers/... as part of the factory...